snuba/datasets/storages/factory.py

```diff
@@ -130,6 +137,7 @@ def __initialize(self) -> None:
                 ]
             },
             **(self._dev_writable_storages if settings.ENABLE_DEV_FEATURES else {}),
+            **self._config_built_writable_storages,
```

Comment:
instead of creating this big dictionary, why not just remove it and have `get_writable_storage_keys` check if the storage is writeable?

Reply:
You'd have to modify `get_writable_storage_keys` to now build a list from a subset of `_all_storages`, and it seems `_all_storages` isn't really _all_ storages since it doesn't contain the config built storages. So one way or another, you have to add these storages to a dict.

Maybe the readable storages should be added as well and then the `get_writable_storage_keys` function could be modified to filter from `_all_storages`

Reply:
> Maybe the readable storages should be added as well and then the get_writable_storage_keys function could be modified to filter from _all_storages

Yeah that's what I mean, we should load all the storages into one place and these convenience functions should just be filters

Reply:
Not sure if I understand the suggestion correctly, but ideally, all processing is done only once during initialization as opposed to every time the method is called.

Reply:
>  but ideally, all processing is done only once during initialization as opposed to every time the method is called.

Why is that ideal?

Reply:
If by processing you mean having to filter the dictionary every time the getter is called, we are already doing a bunch of processing to convert the dict keys to a list before returning. You could prebuild a big all_storages dict and then build small lists of storage keys on init and then just return them in the getter?

Reply:
I think if there we were filtering on numerous parameters, it would be nonsensical to maintain numerous dictionaries which are subsets one set. However, if what we really need is two sets (assuming all storages have been migrated to config), one for all storages and one for only writable storages, then I think two dictionaries is reasonable to maintain. It doesn't cost a lot of memory, but saves us from filtering a dictionary every time the factory method is called.

Reply:
The only reason these `get_<something>_storage_keys` function exist is to provide command line options. They're only called once (if at all). They iterate through a dictionary that's at most 30 elements long, it's a few microseconds of impact.

Meanwhile, look at how many dictionaries are in this init method:

https://github.com/getsentry/snuba/blob/251cad3ab8b1ea648e6599da6eb443859f13ac93/snuba/datasets/storages/factory.py#L24-L34

Those are all extra pieces of state that live on the factory class (and thus globally). Does that state change? Are they serving a useful purpose?

Reply:
Sure, I've added the readable storages to `self._all_storages` and update the methods to filter from that main dictionary instead. Let me know your thoughts on the approach. 

FWIW, I'm not familiar with the context behind `_dev_writable_storages`, `_dev_cdc_storages`, `_non_writable_storages`, and `_dev_non_writable_storages`, but it seems like they are all empty and can be removed which would make that list much less cluttered.

Reply:
Approach is good, let's clean up the __init__ and remove the unnecessary state. Otherwise great job!